dfusion_rust_core/src/models/solution2.rs

```diff
@@ -0,0 +1,204 @@
+use super::*;
+
+use log::info;
+
+use byteorder::{BigEndian, ByteOrder};
+use std::collections::HashMap;
+use std::iter::once;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Solution2 {
+    // token_id => price
+    pub prices: HashMap<u16, u128>,
+    pub executed_buy_amounts: Vec<u128>,
+    pub executed_sell_amounts: Vec<u128>,
+}
+
+impl Solution2 {
+    pub fn trivial(num_orders: usize) -> Self {
+        Solution2 {
+            prices: HashMap::new(),
+            executed_buy_amounts: vec![0; num_orders],
+            executed_sell_amounts: vec![0; num_orders],
+        }
+    }
+
+    // TODO - Why does this have to be a reference?
```

Comment:
As in "Why does my IDE insist the return value be a reference here?"

Reply:
Because the HashMap owns all the keys, so unless you remove all of the keys, you have to return references to the keys instead. You should be able to return `.max().copied()` here as `u16` implements `Copy`.

Reply:
Or even just *(self.prices.keys().max())?

Reply:
Is one better than the other? I guess I wouldn't have to dereference everywhere.

Reply:
> Or even just *(self.prices.keys().max())?

Yes, I could do this instead of having a function, but I find the function is more readable... I could also include `*(self.prices.keys().max())` in the function... Essentially this is just dereferencing before returning.

Reply:
Sadly, 

```
type `std::option::Option<&u16>` cannot be dereferenced
```

Reply:
The other option is `self.prices.keys().max().map(|max| *max)`, but that is pretty much what `.copied()` ends up doing.

Reply:
`.copied()` and `.cloned()` are both `Option` methods for turning an `Option<&T>` into an `Option<T>`. They both work the exact same way except they don't work for the same types. Specifically `Option::copied()` only works for `T: Copy`, so types that implement the `Copy` trait. `Option::cloned()` works for `T: Clone` which is a less restrictive trait. Note that `Copy: Clone` so all types that implement `Copy` also implement `Clone`. `Copy` is just used as a marker to indicate that a type can be cloned, and cloning is super cheap so you shouldn't feel about about using it everywhere.